core/src/main/scala/org/apache/spark/network/netty/NettyBlockRpcServer.scala

```diff
@@ -63,11 +83,83 @@ class NettyBlockRpcServer(
         // StorageLevel is serialized as bytes using our JavaSerializer.
         val level: StorageLevel =
           serializer.newInstance().deserialize(ByteBuffer.wrap(uploadBlock.metadata))
-        val data = new NioManagedBuffer(ByteBuffer.wrap(uploadBlock.blockData))
+        val data = new NioManagedBuffer(LargeByteBufferHelper.asLargeByteBuffer(uploadBlock.blockData))
+        logTrace("putting block into our block manager: " + blockManager)
         blockManager.putBlockData(BlockId(uploadBlock.blockId), data, level)
         responseContext.onSuccess(new Array[Byte](0))
+
+      case uploadPartialBock: UploadPartialBlock =>
+        logTrace("received upload partial block: " + uploadPartialBock)
```

Comment:
This is a key component of block replication for blocks > 2gb.  Definitely need feedback on how robust this approach is, if the handling of dropped messages is sane, etc.